src/data_readers/data_reader_csv.cpp

```diff
@@ -97,127 +97,157 @@ csv_reader::~csv_reader() {
+            label_classes.insert(label);
+            m_labels.push_back(label);
+            break;
+          }
+          cur_pos = end_pos + 1;
         }
-        cur_pos = end_pos + 1;
       }
-    }
-    // Possibly extract the response.
-    if (!m_disable_responses) {
-      size_t cur_pos = 0;
-      for (int col = 0; col < m_num_cols; ++col) {
-        size_t end_pos = line.find_first_of(m_separator, cur_pos);
-        if (col == m_response_col) {
-          DataType response = m_response_transform(
-            line.substr(cur_pos, end_pos - cur_pos));
-          m_responses.push_back(response);
-          break;
+      // Possibly extract the response.
+      if (!m_disable_responses) {
+        size_t cur_pos = 0;
+        for (int col = 0; col < m_num_cols; ++col) {
+          size_t end_pos = line.find_first_of(m_separator, cur_pos);
+          if (col == m_response_col) {
+            DataType response = m_response_transform(
+              line.substr(cur_pos, end_pos - cur_pos));
+            m_responses.push_back(response);
+            break;
+          }
+          cur_pos = end_pos + 1;
         }
-        cur_pos = end_pos + 1;
       }
     }
-  }
-  if (!ifs.eof()) {
-     //If we didn't get to EOF, something went wrong.
-    //throw lbann_exception(
-      //"csv_reader: did not reach EOF");
-  }
-  if (!m_disable_labels) {
-    // Do some simple validation checks on the classes.
-    // Ensure the elements begin with 0, and there are no gaps.
-    auto minmax = std::minmax_element(label_classes.begin(), label_classes.end());
-    if (*minmax.first != 0) {
+
+    if (!ifs.eof()) {
+       //If we didn't get to EOF, something went wrong.
       throw lbann_exception(
-        "csv_reader: classes are not indexed from 0");
+        "csv_reader: did not reach EOF");
     }
-    if (*minmax.second != (int) label_classes.size() - 1) {
-      throw lbann_exception(
-        "csv_reader: label classes are not contiguous");
+    if (!m_disable_labels) {
+      // Do some simple validation checks on the classes.
+      // Ensure the elements begin with 0, and there are no gaps.
+      auto minmax = std::minmax_element(label_classes.begin(), label_classes.end());
+      if (*minmax.first != 0) {
+        throw lbann_exception(
+          "csv_reader: classes are not indexed from 0");
+      }
+      if (*minmax.second != (int) label_classes.size() - 1) {
+        throw lbann_exception(
+          "csv_reader: label classes are not contiguous");
+      }
+      m_num_labels = label_classes.size();
     }
-    m_num_labels = label_classes.size();
+    ifs.clear();
+  } // if (!rank)
+
+  MPI_Bcast(&m_num_cols, 1, MPI_INT, 0, world_comm); 
+  m_label_col = m_num_cols - 1;
+
+  //bcast the index vector
+  int index_size = m_index.size(); 
+  MPI_Bcast(&index_size, 1, MPI_INT, 0, world_comm); 
+  m_index.resize(index_size);
+  MPI_Bcast(m_index.data(), index_size*sizeof(std::streampos), MPI_BYTE, 0, world_comm); 
```

Comment:
I think it would be cleaner to add a scalar-array version of `broadcast` to `lbann_comm` than to use `MPI_Bcast` directly.